### DISCUSSION TITLE
Absolute Persistence a.k.a The Absolute Backdoor

### DISCUSSION THREAD

**Discussion by johannbg** [fwupd/fwupd#3808] [Ideas]
**Is your feature request related to a problem? Please describe.**
Noticed when I was updating the firmware for a security vulnerability via cab files downloaded from the Lenovo side since update was not availabe or being picked up by fwupd yet.

Lenovo provides two cab files for bios update, one update for the `Absolute Persistence Module` if it's enabled and another one for `Non-Absolute-Persistant-Module` encase it has been disabled. 

Now my request is somewhat two fold on one hand the output for the `System Firmware` section and other relevant places should indicate if the system firmware supports absolute persistance and if it's enabled since currently there is no indicator if it's available and enabled/disabled and when you run `fwupdmgr get-details` on both of the provide cab files from Lenovo one get's listed as `Unknown Device` while the other show the `System Firmware` section.

The second part of the RFE is to provide end users the ability to permanently disable the Absolute backdoor, which involves 
some command that offers to disable it and displays some descriptive text to users along the lines of it 
`Once you disable the absolute persistance behaviour it can never be turned back on`
followed by some generic steps that involved 
going into the BIOS Setup,find the Absolute Persistence® Module knob ( should be under security ) 
Select Permanently Disabled. then the usually F10 to Save and Exit.

Then the disable command would valide the above steps ( if doable ) or ask for confirmation from the user that this has been done and then install/replacing the existing firmware with the the non-abolute-persistance firmware ( presumably both firmware versions are being stored by vendors in the firmware database )

A side note which it's better to mention since it's not documented anywhere afaik so you may or may not be aware of it but in the case of Lenovo the command will need to extract the GUID string from firmware.metainfo.xml contained with in the cab file `<firmware type="flashed">THE GUID STRING</firmware>` and use that string instead of the existing GUID one when replacing the non persistant one.

**Comment by hughsie** [comment]
> Lenovo provides two cab files for bios update

I didn't know this. @mrhpearson is this news to you as well? I don't know why a different capsule would be required. I assume the two ESRT tables use different GUIDs for the different types? What happens if you flash the wrong one to the hardware?

> The second part of the RFE is to provide end users the ability to permanently disable the Absolute backdoor

Hmm, providing a 1-way action like this seems a little scary for fwupd. It's probably also out of scope being honest.

**Reply by mrhpearson** [reply]
News to me....
@johannbg - which platform are you seeing this on? Can you point me at the FW and I'll have a look.
Previously for thunderbolt FW updates we had two separate FW versions depending on which mode you were operating in. Sounds like the same here but I'd like to see it in action

I'll have to check on Absolute Persistence - but if there is a knob to control it from Linux then you should do it via the WMI attributes not via fwupd. I have to admit I didn't really follow all the details after that....and got a bit lost. My brain has given up already which doesn't bode well for a Monday :(

**Reply by johannbg** [reply]
@mrhpearson.
 
The particular hw manufacturer I'm referring to is Lenovo and the referrence link I'll be providing are spesific to this laptop I'm typing on ( X1 Carbon Gen6 ) but the broader discussion is applicable to all device manufactures that have partner with [Absolute](https://www.absolute.com/partners/device-manufacturers/). 

When purchasing hardware from Lenovo regular customers have the Absolute Persistence Module version of the bios firmware included by default ( This can be seen under spesific section when custom building a PC on Lenovo's website called `Absolute BIOS Selection` it's included with no option to exclude it ) while mass ordering customers or those that have spesific contracts with Lenovo can request having the Non Absolute Persistence Module version of the bios firmware installed instead.

The `Absolute Persistence Module version` of the bios firmware from Lenovo has a version string that ends with an `W` while the `Non-Absolute Persistence Module version` of the bios firmware from Lenovo has a version string ending in `P`.

Now Lenovo always provides those two bios firmwares version as files part of the BIOS Update Utility for Windows/Linux available on Lenovo's website under --> Drivers & Software --> Bios/UEFI --> BIOS Update Utility ( Windows or Linux ) for the particular hw in question.
 
An example the BIOS Update Utility Linux firmware files for this Carbon Gen6 can be found [here](https://download.lenovo.com/pccbbs/mobiles/n23ul31w.zip) and contains the .cab files for both version of the firmware and the ReadMe which clearly mentions both the firmwares ( Absolute and Non-Absolute ) update [here](https://download.lenovo.com/pccbbs/mobiles/n23ul31w.txt)

For Lenovo to replace the Absolute firmware version of the bios with an Non-Absolute firmware version users will have to perform the following steps...

1. Open BIOS Setup.
2. Move to Security -> Absolute Persistence® Module.
3. Select Permanently Disabled. Then, press F10 to Save and Exit.

To address one of @hughsie concerns of providing a 1-way action of installing the non-absolute version of the bios, the non-absolute bios version cannot be installed unless the user has first permanently disabled the absolute persistence module in the bios. In otherwords this is not an action that can be accidently performed.

4.  Extract the cabininet file ( N23ET78P.cab ). 
5.  Extract the GUID string from the `<firmware type="flashed">$GUID</firmware>` line in the firmware.metainfo.xml file
     ( 99d47dfe-68fc-4a6f-b541-32e003c886fb )
6. run sudo fwupdate -a $GUID-STRING-FROM-FIRMWARE-METAINFO-XML-FILE firmware.bin 
7.  ( sudo fwupdate -a 99d47dfe-68fc-4a6f-b541-32e003c886fb firmware.bin )
8. Reboot

Now at this point I would highlight the fact that the GUID string found in the `Absolute Persistence Module version` of the bios firmware update xml file (  found in N23ET78W.cab metainfo xml file ) would match the output from running `fwupdate -l` but `fwupdate -l` decide to crap itself with...
`
(fwupdate:4645): FuPluginUefiCapsule-WARNING **: 21:12:25.818: SMBIOS BIOS Characteristics Extension Byte 2 is invalid -- UEFI Specification is unsupported, but /sys/firmware/efi exists: SMBIOS not supported
Segmentation fault (core dumped)`

Yay :smile:

**Reply by johannbg** [reply]
Note for those that might be reading the above instructions here and intent to follow them at their own risk :sweat_smile: .
In older Bios firmwares, the Absolute module is called Computrace and the options for it might have some name changes and be presented in a different order but the core of the above instructions are the same and the module needs to be permanently disabled for the bios firmware to be replaced.

**Reply by superm1** [reply]
> (fwupdate:4645): FuPluginUefiCapsule-WARNING **: 21:12:25.818: SMBIOS BIOS Characteristics Extension Byte 2 is invalid -- UEFI Specification is unsupported, but /sys/firmware/efi exists: SMBIOS not supported Segmentation fault (core dumped)

If you can get a `gdb` traceback for this problem and get it into an issue, we should probably fix the segfault.

**Comment by johannbg** [comment]
Arguably this is the exact tool to do ( + you already can, just not in very userfriendly manner ) and the absolute concept is massive security risk for end users.

**Reply by superm1** [reply]
>Hmm, providing a 1-way action like this seems a little scary for fwupd. It's probably also out of scope being honest.
>Arguably this is the exact tool to do ( + you already can, just not in very userfriendly manner ) and the absolute concept is massive security risk for end users.

Does Lenovo allow doing this from software?  I would be really surprised if it's allowed outside of the firmware.  IOW it's an interactivity check that you can only access by actually entering firmware setup.

**Comment by hughsie** [comment]
> and the absolute concept is massive security risk for end users

From a personal point of view, I completely agree. Unfortunately, some people buying the hardware actually *require* it.

**Comment by johannbg** [comment]
Yes corporations,governments and militaries  might want to have this enabled or disabled  depending on the case but the benefit for the regular consumer is little to none thou they are let to believe there is.

Anyways no need for us to engage in a dialog of the usefulness of this certain industry feature and lets just focus on the user experience for those that dont want or need it. 

Lenovo now provides an opt out from this being installed by default on every purchase and provides 2 firmware files, one that includes the backdoor and surveillance software and another one that exclude it.

**Comment by hughsie** [comment]
> one that includes the backdoor and surveillance software and another one that exclude it

Technically we can support this, we have `switch-branch` which can retarget a machine firmware stream, e.g. switching from AMI to coreboot and vice-versa. The question then becomes if Lenovo *wants* to support it.

**Reply by superm1** [reply]
>Technically we can support this, we have switch-branch which can retarget a machine firmware stream, e.g. switching from AMI to coreboot and vice-versa. The question then becomes if Lenovo wants to support it.

If Lenovo _wants_ to support it we would need to come up with a way to associate the two different firmware together and detect which one you're on in a scalable fashion.  I really don't think that quirks that GUID1 is this and GUID2 is that are the right way here - it would probably need to be something that a new Lenovo specific plugin would need to query and set the branch for the device.

**Reply by Classicelevation** [reply]
Hi there my laptop is 5530 12gen I was unlucky to purchase it in refurbished shop deep in town And when I got home to make updates it locked me out with device freeze message

Asking me to contact mastercard my brother passed that one but now everyday I close the like rcpnet file when I put the pc on

Is there anything I can do on this dell latitude 5330 to enable the interface desable

In the bios because now it's active and the interface is grayed out I can't access the other parts

**Comment by Classicelevation** [comment]
Hi there my laptop is 5530 12gen I was unlucky to purchase it in refurbished shop deep in town And when I got home to make updates it locked me out with device freeze message

Asking me to contact mastercard my brother passed that one but now everyday I close the like rcpnet file when I put the pc on

Is there anything I can do on this dell latitude 5330 to enable the interface desable 

In the bios because now it's active and the interface is grayed out I can't access the other parts

**Reply by jaime8505** [reply]
yo tengo un software que lo bloquea desde cualquier windows creado por mi es tuyo por 70 usd el unico completo solo por windows este es mi WhatsApp +3005048853 contactame lo del bios es un riego reprográmalo en esos equipos ya que computrace tiene proteccion para el que se atreva y pierda el equipo
